src/Val.cc

```diff
@@ -1420,6 +1421,115 @@ static void find_nested_record_types(const TypePtr& t, std::set<RecordType*>* fo
     }
 }
 
+// Support class for returning multiple values from a table[pattern]
+// when indexed with a string.
+class detail::TablePatternMatcher {
+public:
+    TablePatternMatcher(const TableVal* _tbl, TypePtr _yield) : tbl(_tbl) {
+        vtype = make_intrusive<VectorType>(std::move(_yield));
+    }
+
+    void Insert(ValPtr pat, ValPtr yield) { Clear(); }
+    void Remove(ValPtr pat) { Clear(); }
+
+    void Clear() { matcher.reset(); }
+
+    VectorValPtr Lookup(const StringValPtr& s);
+
+    // Delegate to matcher->MatchAll().
+    bool MatchAll(const StringValPtr& s);
+
+    void GetStats(detail::DFA_State_Cache_Stats* stats) const {
+        if ( matcher && matcher->DFA() )
+            matcher->DFA()->Cache()->GetStats(stats);
+        else
+            *stats = {0};
+    };
+
+private:
+    void Build();
+
+    const TableVal* tbl;
+    VectorTypePtr vtype;
+
+    // If matcher is nil then we know we need to build it. This gives
+    // us an easy way to cache matchers in the common case that these
+    // sorts of tables don't change their elements very often (indeed,
+    // they'll frequently be constructed just once), and also keeps us
+    // from having to re-build the matcher on every insert/delete in
+    // the common case that a whole bunch of those are done in a single
+    // batch.
+    std::unique_ptr<detail::Specific_RE_Matcher> matcher = nullptr;
+
+    // Maps matcher values to corresponding yields. When building the
+    // matcher we insert a nil at the head to accommodate how
+    // disjunctive matchers use numbering starting at 1 rather than 0.
+    std::vector<ValPtr> matcher_yields;
```

Comment:
So this is the keeping a `Val*` to every yield value in the table so that we can then index into it when we get a match? Looks a bit unfortunate that that additional state is necessary, both memory-wise (though not too bad) and also maintainance-wise (need to rebuild it on changes). However, I also don't see any better approach right now, so ok I guess.

Reply:
IIRC it's just a ValPtr to an existing `Val*` that the table has in its dictionary anyway.

Reply:
To me sounded like templating the AcceptSet/Index type. Then we could associate ValPtr/Val* directly with matching expressions. I'm resolving - this could be interesting if we find more use cases though.